plugin/plugin.go

```diff
@@ -363,6 +373,231 @@ func RegisterShutdown(name string, s Shutter) error {
 	return nil
 }
 
+// Configuration defines the interface for complex_configure callbacks, i.e. Go
+// structs that can hold and validate a configuration
+type Configuration interface {
+	Validate() error
+}
+
+var config_target *Configuration
+var config_targets map[string]*Configuration
+
+var configured map[string]chan struct{}
+
+// getOconfigChildren takes a C.oconfig_item_t and returns an array of Go
+// pointers to its child items. Getting the child pointers requires a little
+// more arithmetic than in C as we can't just increment a pointer to get the
+// next array item
+func getOconfigChildren(oconfig *C.oconfig_item_t) (output []*C.oconfig_item_t) {
+	start := unsafe.Pointer(oconfig.children)
+	size := unsafe.Sizeof(*oconfig.children)
+	for i := 0; i < int(oconfig.children_num); i++ {
+		child := (*C.oconfig_item_t)(unsafe.Pointer(uintptr(start) + size*uintptr(i)))
+		output = append(output, child)
+	}
+	return
+}
+
+// checkMatchable returns an error if the provided source cannot be copied to
+// the target, or nil if it's OK
+func checkMatchable(oconfig *C.oconfig_item_t, target reflect.Value) error {
```

Comment:
The idea driving the two-step process is to try to distinguish between plugin errors that need to be fixed by the developer and config errors that need to be resolved by the user. To that end the checkMatchable function is intended to inspect the provided struct and strictly enforce a few simple rules (only valid types, must be settable fields). This lets us throw an error even if the actual configuration is valid to alert the developer to a potential issue with their struct.

My thinking here is that in Go I'm used to being warned by the compiler if I have type errors, but with the use of Reflect the compiler will let everything through. Having said that I'm happy to defer to your experience as to what would be best for a plugin developer :).